### DISCUSSION TITLE
applicationStorageDirectory keeps reverting to earlier contents

### DISCUSSION THREAD

**Discussion by perholmes** [airsdk/Adobe-Runtime-Support#2287] [General]
I'm facing a bizarre issue, that applicationStorageDirectory keeps reverting to its state from about a week ago every time the app is installed via the IDE. Furthermore, some of the folders have a modification date of Tue Jan 20 1970, 08:15:45 GMT+0100, which isn't even exactly Unix Epoch. One file has specific contents from a week ago, with the code to write that JSON long gone.

I fear the implications of this, and whether applicationStorageDirectory is actually a safe place for user data in this new world of Android 11 and AIR 50.

Sure, it might be a development anomaly, but it's awfully specific, and it indicates that there's a shadow applicationStorageDirectory somewhere that it's possible to get returned to. What if it's possible for users to get their data returned to an earlier state in a live app? What triggers this rollback?

Does anyone have any clue what might be happening here? The only thing I can guess is that installing a debug version over USB is different from running from the Play Store, and it has its own storage directory. But still, what froze the basic contents to a week ago? Why isn't the directory empty after a clean uninstall and reinstall? I rely on the absence of certain files to know to migrate data.

Any ideas or speculation would be much appreciated!

**Comment by perholmes** [comment]
Having spent a full day on this, it's truly bizarre. No matter what is done to the applicationStorageDirectory, it does a Groundhog Day back to a few days ago every time the app is installed from the IDE.

Here's a list of what made no difference:

- Physically uninstalling the app.
- Uninstall it from the beta program via Google Play, in case it interfered with IDE installation.
- Clearing app cache and/or data from Android settings.
- Restarting phone.
- Putting files into the directory. They disappear on next launch.
- Recursively deleting everything in the directory. It's back on next launch.

Where is this shadow applicationStorageDirectory coming from? Everything I've done to write and delete files works fine in session -- the files are created/deleted as expected -- but on next launch, it's 3 days ago again. That indicates to me that it's not merely some kind of symlink mapping changing. The shadow settings are literally being copied into /data/user/0/bundleid/bundleid/Local Store/... on every launch.

**Reply by ajwfrost** [reply]
Seems really weird..

Is this only happening when you're installing the app from the IDE? What if you create the app package (APK file or AAB) from the IDE and then install it manually, clean up, and install the same package again?

I don't suppose you have a fresh phone you can try it on, just in case something's gone really weird with Android on one device?

**Reply by perholmes** [reply]
Thanks for getting back. Those would be the two next logical steps to try, which I'll do tomorrow. I'll also try to do an internal test build via Google Play. I only need to come up with a way to capture the logging once it's running outside of debug.

I'll paste the directory listing here of what keeps coming back:

`[trace] ----- Application Storage Directory Listing -----`
`[trace] (DIR) /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/Shot Designer Scenes (Wed Nov 9 2022, 17:52:05 GMT+0100) `
`[trace] (DIR) /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/Shot Designer Scenes/Imported (Tue Jan 20 1970, 08:15:45 GMT+0100) `
`[trace] (DIR) /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/Shot Designer Scenes/My Templates (Tue Jan 20 1970, 08:15:45 GMT+0100) `
`[trace] (DIR) /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/Shot Designer Scenes/Shared With Me (Tue Jan 20 1970, 08:15:45 GMT+0100) `
`[trace] (DIR) /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/Shot Designer Scenes/My Scenes (Tue Jan 20 1970, 08:15:45 GMT+0100) `
`[trace] /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/androidmigration.dat (Tue Jan 20 1970, 08:15:45 GMT+0100) `
`[trace] /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/ShotDesignerPreferences.xml (Tue Jan 20 1970, 08:15:48 GMT+0100) `
`[trace] /data/user/0/air.us.hollywoodcamerawork.shotdesigner/us.hollywoodcamerawork.shotdesigner/Local Store/StartupCrashDetection.txt (Wed Nov 9 2022, 17:52:10 GMT+0100) `
`[trace] ----- Contents of androidmigration.dat -----`
`[trace] true`
`[trace] ----- Done -----`

What's especially telling is that the androidmigration.dat file was written as a "true" or "false" string 3-4 days ago. But now the migration logic is just based on the presence of the file, which then has "Android migration complete" written into it. If I force a migration, this directory listing prints the right value. But if I just run the app, this directory is back. If I delete it, it's back. If I fill it up with files, they're gone on the next install (via IDE).

So definitely yes, we need to eliminate the IDE as a cause.

**Reply by ajwfrost** [reply]
Do you have a File Manager app on the phone that you can use to independently check whether those folders do exist? I'm just wondering whether there's actually just a bug in AIR that's somehow caching a value and thinking that these folders are there even if they're not.. (there shouldn't be anything like that as far as I'm aware, but it does seem suspicious that if you put files into this location, they don't actually persist...)

**Comment by perholmes** [comment]
Unfortunately, no file manager has access to applicationStorageDirectory on Android 11+ unless the device is rooted. All file manager devs are complaining. The file manager only has access to its own directory in /data/

But the files really do exist. The androidmigration.dat file literally has contents as the code wrote them a week ago.

I'll be doing the next tests today, clean install without IDE/USB involved, as well as a test build distributed via Google Play.

**Comment by marchbold** [comment]
It's not getting restored by the android auto backup process is it? 

https://developer.android.com/guide/topics/data/autobackup 

You can try adding `android:allowBackup="false"` to your manifest additions `application` tag:

```xml
<application android:allowBackup="false" ... >
```

**Comment by perholmes** [comment]
Michael, you nailed it. I truly had no idea this feature existed, and I somehow never confronted that this must have been happening for half a decade. Only now that I cared about the specific state of the files did I notice.

So basically, Android has a backup feature that backs up this directory to Google Drive, which happens when the device is idle, every 24 hours. Apparently, it's also keeping a local cache of this, so that repeated installs via the IDE immediately hand a populated data directory to the app. They specifically say that AutoBackup works regardless of how the app was installed.

What's perhaps strange is that it was stuck on data from a week ago, so I didn't get a backup in 7 days. But it doesn't matter. This seems like a good feature for Android to implement, I just have to test without it.

Amazing. You get into such a cross-platform headspace that you don't really understand the individual platforms well enough, which is a problem.

**Reply by marchbold** [reply]
Yeah it's a great feature they implemented, but can catch you out if you aren't expecting it. 

Definitely takes a lot of time to know all the intricacies of each of the platforms, but at least you don't have to recode your app as well ;)

**Reply by marchbold** [reply]
Also from memory, I think you could test without it by logging out all Google accounts on the device.

**Reply by MalacTheLittle** [reply]
I have never been adding `<application android:allowBackup="false">` to app descriptor xml and I have never had the issue like @perholmes did although I'm using `applicationStorageDirectory` quite often. But when saving files I'm putting `File.preventBackup` to `true` by default. 
Could that be a reason I've been avoiding this issue?

**Reply by perholmes** [reply]
That sounds quite possible that always  setting that flag (with perfect discipline) would sidestep the issue. But I think it might be worthwhile to still check up on the docs and verify that this File flag is still being honored on Android 12+, because they have all new rules for controlling what gets backed up.

**Comment by perholmes** [comment]
One thing to also become aware of is that Android wants backup rules as XML for what to back up, and these are fundamentally different from Android 12 and pre-Android 12. I could have gone another 5 years not knowing this existed, so it's good that I found out.

**Reply by teotigraphix** [reply]
Android dev, a thugs life. ;-) Just sharing the pain with fellow devs.

**Comment by MalacTheLittle** [comment]
> But I think it might be worthwhile to still check up on the docs and verify that this File flag is still being honored on Android 12+, because they have all new rules for controlling what gets backed up.

That's what I'm afraid of now. 
@ajwfrost Maybe you can confirm if `File.preventBackup` flag has same effect as descriptor `<application android:allowBackup="false">` tag?
